File: examples/python/run_sync_model_memory.py
#
# Copyright (C) 2018- DEEPX Ltd.
# All rights reserved.
#
# This software is the property of DEEPX and is provided exclusively to customers 
# who are supplied with DEEPX NPU (Neural Processing Unit). 
# Unauthorized sharing or usage is strictly prohibited by law.
#

import numpy as np
import os
import argparse
import time
from dx_engine import InferenceEngine
from logger import Logger, LogLevel


def parse_args():
    parser = argparse.ArgumentParser(description="Run synchronous model memory inference")
    parser.add_argument("--model", "-m", type=str, required=True, help="Path to model file (.dxnn)")
    parser.add_argument("--loops", "-l", type=int, default=1, help="Number of inference loops (default: 1)")
    parser.add_argument("--verbose", "-v", action="store_true", default=False, help="Enable debug logging")
    args = parser.parse_args()

    if not os.path.exists(args.model):
        parser.error(f"Model path '{args.model}' does not exist.")
    
    if args.verbose:
        logger = Logger()
        logger.set_level(LogLevel.DEBUG)
    
    return args


if __name__ == "__main__":
    args = parse_args()
    logger = Logger()
        
    logger.info(f"Start run_sync_model test for model: {args.model}")
    
    try:
        # Load model file into buffer
        logger.info(f"Loading model file: {args.model}")
        with open(args.model, 'rb') as f:
            model_buffer = np.frombuffer(f.read(), dtype=np.uint8)
        logger.info(f"Model loaded into buffer: {len(model_buffer)} bytes")
        
        # create inference engine instance with model buffer
        logger.info("Creating InferenceEngine from model buffer")
        with InferenceEngine.from_buffer(model_buffer) as ie:

            # NOTE: np.zeros() uses COW zero pages — all virtual pages share one
            # physical page. PCIe DMA driver's get_user_pages() then sees duplicate
            # physical pages in the SG list and fails with EFAULT.
            # np.empty() + explicit fill forces unique physical page allocation.
            _buf = np.empty(ie.get_input_size(), dtype=np.uint8)
            _buf.fill(0)
            input = [_buf]

            start = time.perf_counter()
            # inference loop
            for i in range(args.loops):

                # inference synchronously 
                # use only one npu core 
                # ownership of the outputs is transferred to the user 
                outputs = ie.run(input)

                # post processing 
                #postProcessing(outputs)
                logger.debug(f"Inference outputs {i}")
                
            end = time.perf_counter()
            total_time_ms = (end -start) * 1000
            avg_latency = total_time_ms / args.loops
            fps = 1000.0/ avg_latency if avg_latency > 0 else 0.0
            
            logger.info("-----------------------------------")
            logger.info(f"Total Time: {total_time_ms:.3f} ms")
            logger.info(f"Average Latency: {avg_latency:.3f} ms")
            logger.info(f"FPS: {fps:.2f} frame/sec")
            logger.info("Success")
            logger.info("-----------------------------------")
                
    except Exception as e:
        logger.error(f"Exception: {str(e)}")
        exit(-1)
        
    exit(0)

File: examples/cpp/run_sync_model_memory/run_sync_model_memory.cpp
/*
 * Copyright (C) 2018- DEEPX Ltd.
 * All rights reserved.
 *
 * This software is the property of DEEPX and is provided exclusively to customers
 * who are supplied with DEEPX NPU (Neural Processing Unit).
 * Unauthorized sharing or usage is strictly prohibited by law.
 */

#include "dxrt/dxrt_api.h"
#include "dxrt/extern/cxxopts.hpp"
#include "../include/logger.h"
#include <string>
#include <iostream>
#include <fstream>
#include <vector>


int main(int argc, char* argv[])
{
    std::string model_path;
    int loop_count;
    bool verbose;

    auto &log = dxrt::Logger::GetInstance();

    cxxopts::Options options("run_sync_model_memory", "Run synchronous model memory inference");
    options.add_options()
        ("m,model", "Path to model file (.dxnn)", cxxopts::value<std::string>(model_path))
        ("l,loops", "Number of inference loops", cxxopts::value<int>(loop_count)->default_value("1"))
        ("v,verbose", "Enable verbose/debug logging", cxxopts::value<bool>(verbose)->default_value("false"))
        ("h,help", "Print usage");

    try
    {
        auto result = options.parse(argc, argv);

        if (result.count("help") || !result.count("model"))
        {
            std::cout << options.help() << std::endl;
            return result.count("help") ? 0 : -1;
        }

        if (verbose) {
            log.SetLevel(dxrt::Logger::Level::LOGLEVEL_DEBUG);
        }
    }
    catch (const std::exception& e)
    {
        log.Error(std::string("Error parsing arguments: ") + e.what());
        std::cout << options.help() << std::endl;
        return -1;
    }

    log.Info("Start run_sync_model_memory test for model: " + model_path);

    try
    {
        // Load model file into memory first
        std::ifstream model_file(model_path, std::ios::binary | std::ios::ate);
        if (!model_file.is_open()) {
            log.Error("Failed to open model file: " + model_path);
            return -1;
        }

        std::streamsize model_size = model_file.tellg();
        model_file.seekg(0, std::ios::beg);

        std::vector<uint8_t> model_buffer(model_size);
        if (!model_file.read(static_cast<char*>(static_cast<void*>(model_buffer.data())), model_size)) {
            log.Error("Failed to read model file: " + model_path);
            return -1;
        }
        model_file.close();

        log.Info("Model loaded into memory, size: " + std::to_string(model_size) + " bytes");

        // create inference engine instance with model from memory
        dxrt::InferenceEngine ie(model_buffer.data(), model_buffer.size());

        // create temporary input buffer for example
        std::vector<uint8_t> inputPtr(ie.GetInputSize(), 0);

        auto start = std::chrono::high_resolution_clock::now();

        // inference loop
        for(int i = 0; i < loop_count; ++i)
        {

            // inference synchronously, use one npu core
            auto outputs = ie.Run(inputPtr.data());

            log.Debug("Inference outputs (" + std::to_string(i) + ")");

            // now there is no post processing
            (void)outputs;

        }

        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double, std::milli> duration = end - start;

        double total_time = duration.count();
        double avg_latency = total_time / static_cast<double>(loop_count);
        double fps = 1000.0 / avg_latency;

        log.Info("-----------------------------------");
        log.Info("Total Time: " + std::to_string(total_time) + " ms");
        log.Info("Average Latency: " + std::to_string(avg_latency) + " ms");
        log.Info("FPS: " + std::to_string(fps) + " frame/sec");
        log.Info("Success");
        log.Info("-----------------------------------");
    }
    catch (const dxrt::Exception& e)
    {
        log.Error(std::string("dxrt::Exception: ") + e.what());
        return -1;
    }
    catch (const std::exception& e)
    {
        log.Error(std::string("std::exception: ") + e.what());
        return -1;
    }
    catch(...)
    {
        log.Error("Exception");
        return -1;
    }

    return 0;
}
